File: tobas_gcs/src/tobas_gcs/apps/mission_planner/add_command_dialog.py
from PyQt5.QtCore import pyqtSlot
from PyQt5.QtWidgets import QWidget, QDialog, QPushButton, QVBoxLayout, QHBoxLayout

from tobas_rqt_tools.widgets import ListWidget

from .structs import Commands


class AddCommandDialog(QDialog):
    def __init__(self, parent: QWidget) -> None:
        super().__init__(parent=parent)

        self.setWindowTitle("Add Command")

        rows = QVBoxLayout()
        self.setLayout(rows)

        self._cmd_list = ListWidget()
        self._cmd_list.addItems(Commands.values())
        self._cmd_list.setSelectionMode(ListWidget.SingleSelection)  # 単一選択
        rows.addWidget(self._cmd_list)

        cols = QHBoxLayout()
        rows.addLayout(cols)

        self._cancel_button = QPushButton("Cancel")
        self._cancel_button.clicked.connect(self.reject)
        cols.addWidget(self._cancel_button)

        self._ok_button = QPushButton("OK")
        self._ok_button.clicked.connect(self._on_ok_clicked)
        cols.addWidget(self._ok_button)

        self._selected_command = ""

    def selected_command(self) -> str:
        return self._selected_command

    @pyqtSlot()
    def _on_ok_clicked(self) -> None:
        selected_item = self._cmd_list.selected_item()
        if selected_item is None:
            return

        self._selected_command = selected_item.text()
        self.accept()

File: tobas_gui/tobas_control_system/src/mission_planner/add_command_dialog.cpp
#include <QPushButton>
#include <QVBoxLayout>
#include <QHBoxLayout>

#include <magic_enum/magic_enum.hpp>

#include "tobas_control_system/mission_planner/add_command_dialog.hpp"

namespace gui
{
namespace gcs
{
AddCommandDialog::AddCommandDialog(QWidget* parent) : super(parent)
{
  setWindowTitle("Add Command");

  command_list_ = new qt::ListWidget();
  command_list_->setSelectionMode(QListWidget::SingleSelection);
  for (const auto& cmd : magic_enum::enum_values<command_t>())
    command_list_->addItem(commandToText(cmd));

  const auto cancel_button = new QPushButton("Cancel");
  const auto ok_button = new QPushButton("OK");
  ok_button->setDefault(true);

  // Layout
  const auto cols = new QHBoxLayout();
  cols->addWidget(cancel_button);
  cols->addWidget(ok_button);

  const auto rows = new QVBoxLayout();
  rows->addWidget(command_list_);
  rows->addLayout(cols);

  setLayout(rows);

  // Connection
  connect(cancel_button, &QPushButton::clicked, this, &self::reject);
  connect(ok_button, &QPushButton::clicked, this, &self::onOkClicked);
}

command_t AddCommandDialog::selectedCommand() const
{
  return selected_command_;
}

void AddCommandDialog::onOkClicked()
{
  const auto selected_item = command_list_->selectedItem();
  if (selected_item == nullptr)
    return;

  selected_command_ = textToCommand(selected_item->text().toUtf8());
  accept();
}
};  // namespace gcs
}  // namespace gui
